File: Java/PilhaEstatica/PilhaEstatica/src/PilhaEstatica.java
public class PilhaEstatica <T> implements Empilhavel{
    private int ponteiroTopo;
    private T[] dados;

    public PilhaEstatica(int tamanho) {
        ponteiroTopo = -1;
        dados = (T[]) new Object[tamanho];
    }

    public PilhaEstatica() {
        this(10);
    }

    @Override
    public void empilhar(Object dado) {
        if (!estaCheia()) {
            ponteiroTopo++;
            dados[ponteiroTopo] = (T) dado;
        } else {
            System.out.println("Pilha Cheia!");
        }
    }

    @Override
    public Object desempilhar() {
        Object retorno = null;

        if (!estaVazia()) {
            retorno = dados[ponteiroTopo];
            ponteiroTopo--;
        } else {
            System.out.println("Pilha Vazia!");
        }

        return retorno;
    }

    @Override
    public Object topo() {
        Object retorno = null;

        if (!estaVazia()) {
            retorno = dados[ponteiroTopo];
        } else {
            System.out.println("Pilha Vazia!");
        }

        return retorno;
    }

    @Override
    public boolean estaCheia() {
        return ponteiroTopo == dados.length - 1;
    }

    @Override
    public boolean estaVazia() {
        return ponteiroTopo == -1;
    }

    @Override
    public String imprimir() {
        String retorno = "[";

        for (int i = ponteiroTopo; ponteiroTopo > -1; i--) {
            retorno += dados[i] + "\n";
        }

        return retorno;
    }
}

File: Kotlin/PilhaEstatica/PilhaEstatica/src/PilhaEstatica.kt
class PilhaEstatica(private val tamanho: Int = 10): Empilhavel {
    private var ponteiroTopo: Int = -1
    private var dados: Array<Any?> = arrayOfNulls(tamanho)

    // metodos 1

    override fun empilhar(dado: Any?) {
        if (!estaCheia()) {
            ponteiroTopo++
            dados[ponteiroTopo] = dado
        } else {
            println("Pilha Cheia!");
        }
    }

    override fun desempilhar(): Any? {
        var retorno: Any? = null

        if (!estaVazia()) {
            retorno = dados[ponteiroTopo]
            ponteiroTopo--
        }

        return retorno
    }

    override fun topo(): Any? {
        var retorno: Any? = null

        if (!estaVazia()) {
            retorno = dados[ponteiroTopo]
        }

        return retorno
    }

    // metodos 2

    override fun estaCheia(): Boolean {
        return (ponteiroTopo == dados.size - 1)
    }

    override fun estaVazia(): Boolean {
        return (ponteiroTopo == -1)
    }

    override fun imprimir(): String {
        var impressao = "["

        for (i in ponteiroTopo downTo 0) {
            if (i == 0) {
                impressao += "${dados[i]}"
            } else {
                impressao  += "${dados[i]}, "
            }
        }

        return "${impressao}]"
    }
}